src/common/arrow/arrow_array_scan.cpp

```diff
@@ -0,0 +1,580 @@
+                StructVector::getFieldVector(outputVector, curType).get(), mask->getChild(curType),
+                curOffset + srcOffset, i + dstOffset, 1);
+            // may be inefficient, since we're only scanning a single value
+            // should probably ask if we support dense unions (ie. is it okay to pack them)
+        }
+    }
+}
+
+void scanArrowArraySparseUnion(const ArrowSchema* schema, const ArrowArray* array,
+    ValueVector* outputVector, ArrowNullMaskTree* mask, int64_t srcOffset, int64_t dstOffset,
+    int64_t cnt) {
+    const int8_t* types = ((const int8_t*)array->buffers[0]) + srcOffset;
+    mask->copyToValueVector(outputVector, dstOffset, cnt);
+    for (int64_t i = 0; i < cnt; i++) {
+        if (!mask->isNull(i)) {
+            int8_t curType = types[i];
+            UnionVector::setTagField(outputVector, curType);
+        }
+    }
+    // it is specified that values that aren't selected in the type buffer
+    // must also be semantically correct. this is why this scanning works.
+    // however, there is possibly room for optimization here.
+    // eg. nulling out unselected children
+    for (int8_t i = 0; i < array->n_children; i++) {
+        ArrowArrayScan::scanArrowArray(schema->children[i], array->children[i],
+            UnionVector::getValVector(outputVector, i), mask->getChild(i), srcOffset, dstOffset,
+            cnt);
+    }
+}
+
+template<typename offsetsT>
+void scanArrowArrayDictionaryEncoded(const ArrowSchema* schema, const ArrowArray* array,
+    ValueVector* outputVector, ArrowNullMaskTree* mask, int64_t srcOffset, int64_t dstOffset,
+    int64_t cnt) {
+
+    const offsetsT* values = ((const offsetsT*)array->buffers[1]) + srcOffset;
+    mask->copyToValueVector(outputVector, dstOffset, cnt);
+    for (int64_t i = 0; i < cnt; i++) {
+        if (!mask->isNull(i)) {
+            auto dictOffseted = (*mask->getDictionary()) + values[i];
+            ArrowArrayScan::scanArrowArray(schema->dictionary, array->dictionary, outputVector,
+                &dictOffseted, values[i] + array->dictionary->offset, i + dstOffset,
+                1); // possibly inefficient?
+        }
+    }
+}
+
+void scanArrowArrayRunEndEncoded(const ArrowSchema* schema, const ArrowArray* array,
+    ValueVector* outputVector, ArrowNullMaskTree* mask, int64_t srcOffset, int64_t dstOffset,
+    int64_t cnt) {
+
+    const ArrowArray* runEndArray = array->children[0];
+    const int32_t* runEndBuffer = (const int32_t*)runEndArray->buffers[1];
+
+    // binary search run end corresponding to srcOffset
+    int64_t runEndIdx = runEndArray->offset;
+    {
+        int64_t L = runEndArray->offset, H = L + runEndArray->length;
+        while (H >= L) {
+            int64_t M = (H + L) >> 1;
+            if (runEndBuffer[M] < srcOffset) {
+                runEndIdx = M;
+                H = M - 1;
+            } else {
+                L = M + 1;
+            }
+        }
+    }
+
+    for (int64_t i = 0; i < cnt; i++) {
```

Comment:
Instead of scanning one value into outputVector at a time, a better way is to loop over each value in runEndBuffer and figure out the length of it, then do the copy. This way can at least avoid redundant `while` loop here.

For this comment "// there is optimization to be made here...", can you write the optimization down here? better document it now, otherwise we might never be able to recall and revisit it.

Reply:
It may be more efficient to copy native values to native values, rather than copying arrow values to kuzu values. I was thinking of copying the value to kuzu, and then copying the kuzu value over and over again to fill the run-end segment. Not sure how much of an improvement this will be.

It's also less efficient to perform one-by-one copying. Calling `fromArrowArray` over and over again incurs the overhead of recursively calling a function, and there may also be more efficient ways to fill run-end segments of primitive values.

Reply:
Also, Run-end encoding is essentially fancy dictionary encoding so looping through the runEndBuffer instead of from 0 to `cnt` still requires one-by-one copying. I think I understand what you're going for, but changing the way the loops work isn't really going to do anything.